### PR TITLE
update NOAA fetch/munge

### PR DESCRIPTION
**Pull request by amsnyder**
Updated code to:
- fetch data month by month due to API request restrictions
- fetch data for all NOAA sties
- add conductivity to list of variables to fetch

I also switched to the mean sea level datum, as recommended by Salme.

### CODE REVIEW

**Comment by ted80810** [comment]
@amsnyder You can go ahead and delete those outputs from S3. I appreciate you asking. I am still checking out the your updates locally. It looks good so far though.

File: 02_munge/src/munge_noaa_nos.py [hunk 1]
```diff
@@ -1,24 +1,120 @@
...
+    write_location = config['write_location']
+    s3_client = utils.prep_write_location(write_location, config['aws_profile'])
+    s3_bucket = config['s3_bucket']
+
+    # get list of raw data files to process, by site
+    with open("01_fetch/fetch_config.yaml", 'r') as stream:
+        station_ids = yaml.safe_load(stream)['fetch_noaa_nos.py']['station_ids']
+    raw_datafiles = {}
+    for station_id in station_ids:
+        raw_datafiles[station_id] = [obj['Key'] for obj in s3_client.list_objects_v2(Bucket=s3_bucket, Prefix=f'01_fetch/out/noaa_nos_{station_id}')['Contents']]
```
Comment (galengorski) [inline 02_munge/src/munge_noaa_nos.py:104]: Does this incorporate a write local option or is writing to S3 the only option now?

**Review by galengorski — COMMENTED** [review]
I added a few comments and suggestions that made the code run on my machine, otherwise looks good!

File: 02_munge/src/munge_noaa_nos.py [hunk 2]
```diff
@@ -1,24 +1,120 @@
...
+    # save pre-processed data
+    data_outfile_csv = os.path.join('.', '02_munge', 'out', f'noaa_nos_{site}.csv')
+    combined_df.to_csv(data_outfile_csv, index=True)
+    
     if write_location == 'S3':
-           print('uploading to s3')
-           s3_client.upload_file(data_outfile_csv, s3_bucket, '02_munge/out/'+os.path.basename(data_outfile_csv))
\ No newline at end of file
+        print('uploading to s3')
+        s3_client.upload_file(data_outfile_csv, s3_bucket, '02_munge/out/'+os.path.basename(data_outfile_csv))
```
Comment (amsnyder) [inline 02_munge/src/munge_noaa_nos.py:84]:
```suggestion
        s3_client.upload_file(data_outfile_csv, s3_bucket, '02_munge/out/'+os.path.basename(data_outfile_csv))
        
    return combined_df
```

File: 02_munge/src/munge_noaa_nos.py [hunk 3]
```diff
@@ -1,24 +1,120 @@
...
+    # determine which QA/QC levels we want to drop
+    qa_to_drop = config['qa_to_drop']
+    # number of measurements required to consider average valid
+    prop_obs_required = config['prop_obs_required']
+    # timestep to aggregate to
+    agg_level = config['agg_level']
+
+    # process raw data files into csv
+    for site, site_raw_datafiles in raw_datafiles.items():
+        process_data_to_csv(site, site_raw_datafiles, qa_to_drop, flags_to_drop_by_var, agg_level, prop_obs_required, read_location, write_location, s3_bucket, s3_client)
```
Comment (amsnyder) [inline 02_munge/src/munge_noaa_nos.py:117]:
```suggestion
        df = process_data_to_csv(site, site_raw_datafiles, qa_to_drop, flags_to_drop_by_var, agg_level, prop_obs_required, read_location, write_location, s3_bucket, s3_client)
        # apply butterworth filter
```

File: 02_munge/src/munge_noaa_nos.py [hunk 4]
```diff
@@ -1,24 +1,123 @@
...
-    #grab each dataset from bucket and combine to a single csv file
-        for obs in prefix_obs:
-            obj = s3_client.get_object(Bucket=s3_bucket, Key=obs)
-            temp = pd.read_csv(obj.get('Body'), index_col=None, header=0, encoding='utf8')
-            prefix_df.append(temp)
-    #sort columns to timestep
-    df = pd.concat(prefix_df, ignore_index=True, sort = True)
-    data_outfile_csv = f'{station_id}_{product}'
-    #save to csv file
+def read_data(raw_datafile, read_location, s3_bucket):
```
Comment (amsnyder) [inline 02_munge/src/munge_noaa_nos.py:8]:
```suggestion
def get_datafile_list(station_ids, read_location, s3_client=None, s3_bucket=None):
    raw_datafiles = {}
    if read_location=='S3':
        for station_id in station_ids:
            raw_datafiles[station_id] = [obj['Key'] for obj in s3_client.list_objects_v2(Bucket=s3_bucket, Prefix=f'01_fetch/out/noaa_nos_{station_id}')['Contents']]
    elif read_location=='local':
        prefix = os.path.join('01_fetch', 'out')
        for station_id in station_ids:
            file_prefix=f'noaa_nos_{station_id}'
            raw_datafiles[station_id] = [os.path.join(prefix, f) for f in os.listdir(prefix) if f.startswith(file_prefix)]
    return raw_datafiles

def read_data(raw_datafile, read_location, s3_bucket):
```

File: 02_munge/src/munge_noaa_nos.py [hunk 5]
```diff
@@ -1,24 +1,123 @@
...
+    write_location = config['write_location']
+    s3_client = utils.prep_write_location(write_location, config['aws_profile'])
+    s3_bucket = config['s3_bucket']
+
+    # get list of raw data files to process, by site
+    with open("01_fetch/fetch_config.yaml", 'r') as stream:
+        station_ids = yaml.safe_load(stream)['fetch_noaa_nos.py']['station_ids']
+    raw_datafiles = {}
+    for station_id in station_ids:
+        raw_datafiles[station_id] = [obj['Key'] for obj in s3_client.list_objects_v2(Bucket=s3_bucket, Prefix=f'01_fetch/out/noaa_nos_{station_id}')['Contents']]
```
Comment (amsnyder) [inline 02_munge/src/munge_noaa_nos.py:118]:
```suggestion
    raw_datafiles = get_datafile_list(station_ids, read_location, s3_client, s3_bucket)
```

File: 02_munge/src/munge_noaa_nos.py (hunk 1 above)
Comment (amsnyder) [reply 02_munge/src/munge_noaa_nos.py]: just updated it to be fully a local read (and not list the files on S3...let me know if you have any trouble with it!

File: 02_munge/src/munge_usgs.py [hunk 6]
```diff
@@ -5,19 +5,6 @@
...
-        expected_measurements = df.groupby([df['datetime'].dt.date]).count().mode()[cols].loc[0]
-        observed_measurements = df.groupby([df['datetime'].dt.date]).count()[cols].loc[:]
-        prop_df = observed_measurements / expected_measurements
-        # calculate averages for timestep
-        df = df.groupby([df['datetime'].dt.date]).mean()
-    # only keep averages where we have enough measurements
-    df.where(prop_df.gt(prop_obs_required), inplace=True)
-    return df
-
 def param_code_to_name(df, params_df):
```
Comment (amsnyder) [inline 02_munge/src/munge_usgs.py:8]:
```suggestion
def get_datafile_list(read_location, s3_client=None, s3_bucket=None):
    raw_datafiles = {}
    if read_location=='S3':
        raw_datafiles = [obj['Key'] for obj in s3_client.list_objects_v2(Bucket=s3_bucket, Prefix='01_fetch/out/usgs_nwis_0')['Contents']]
    elif read_location=='local':
        prefix = os.path.join('01_fetch', 'out')
        file_prefix='usgs_nwis_0'
        raw_datafiles = [os.path.join(prefix, f) for f in os.listdir(prefix) if f.startswith(file_prefix)]
    return raw_datafiles

def param_code_to_name(df, params_df):
```

**Comment by amsnyder** [comment]
@ted80810 - do you think we can merge this and continue to update the butterworth filter in another pull request?

**Comment by galengorski** [comment]
I just pulled this PR into a new branch on my local and tried to run it. It seemed to run fine and produce the correct files, but I got the following error. I am happy to try to track this down and debug, but I am trying to move forward with the analysis and don't want to get bogged down so I'll just leave it here for now and will return later.
munge_error

**Comment by amsnyder** [comment]
I am going to merge this PR, but let's revisit this error

**Pull request merged**